patch/matching.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package patch
+
+import (
+	"fmt"
+	"log"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/microsoft/go-infra/gitcmd"
+)
+
+// MatchCheckRepo checks whether each patch in a series of patches makes actual changes, or if it is
+// equivalent to an existing patch.
+type MatchCheckRepo struct {
+	gitDir                     string
+	existingPatchPathsByHeader map[Header]string
+
+	// LastApplyExistingMatch is the match for the last patch file passed to the Apply method, or ""
+	// if no match was found. The purpose is to provide "git go-patch extract" functionality that
+	// ignores superficial differences like hunk line numbers and index hashes that would only
+	// distract code reviewers.
+	LastApplyExistingMatch string
+}
+
+// NewMatchCheckRepo clones the given submodule to a temp repo and prepares to search the patchesDir
+// for a match to each patch passed to Apply. Returns the created context for this process. Call the
+// context's AttemptDelete method to clean up the temp dir if desired.
+func NewMatchCheckRepo(submodulePath, baseCommit, patchesDir string) (*MatchCheckRepo, error) {
+	m := MatchCheckRepo{}
+	var err error
+	if m.gitDir, err = gitcmd.NewTempCloneRepo(submodulePath); err != nil {
+		return nil, fmt.Errorf("failed to create temp clone of submodule to evaluate patch changes: %v", err)
+	}
+	if err := gitcmd.Run(m.gitDir, "checkout", "-q", baseCommit); err != nil {
+		return nil, fmt.Errorf("failed to check out base commit in temp repo: %v", err)
+	}
+	// Associate each patch's header (identity, as far as this comparison is concerned) with the
+	// patch file's path. This lets us try old and new when we find a new patch's header, later.
+	m.existingPatchPathsByHeader = make(map[Header]string)
+	if err := WalkPatches(patchesDir, func(path string) error {
```

Comment:
Nit: may want to update docs for `WalkPatches` to clarify it's returned in a deterministic fashion. The internal docs indicate this but may help in the summary. Always suspicious of anything that involves the file system.